### ISSUE TITLE
high memory usage

### ISSUE THREAD

**Issue by jamieS95** [JoeyDeVries/LearnOpenGL#197]
![memory](https://user-images.githubusercontent.com/50465426/94349488-a919a500-003c-11eb-9b86-b59a866055b5.png)

**Comment by gameengineer** [comment]
The code in this course isnt meant to be a high efficient game engine. It's for teaching graphics.

**Comment by jamieS95** [comment]
all the more reason to teach them properly.
its better to teach them to write efficiently than teach them bad practices.

**Comment by JoeyDeVries** [comment]
I disagree. Code that is simpler to understand, while not the most efficient, strongly has my preference for educational purposes. Graphics programming is already hard enough as it is, without also having to worry about the most performant code. Of course, code that's easy to understand and efficient at the same time would be best, but that balance isn't always straightforward. 

If there's a bug in the code that's causing high memory usage that's a different problem, if you can figure out what's causing it I'd love to fix that.

**Comment by amengol** [comment]
For starters, I am not sure about the relevance of your picture. I could use Joey's tutorials to load thousands of high poly models and make it perform really bad, then take a similar picture and blindly blame his code. This is a nonsense issue until you show what the problem really is (if any).

That being said. Joey's course is one of the most comprehensive and easy to understand computer graphics materials out there. The fact that it was not written to be high efficient, as said above, is one of the reasons it has helped many people so far, since optimized code tends to be difficult to understand/maintain. Besides, the topic is about OpenGL, not optimization. There in an unlimited number of topics and books available about high efficient code to serve you well though.

**Comment by jamieS95** [comment]
@JoeyDeVries  
Your code is easy to understand to an extent, you have a shader, model, mesh class...
you upload the VBO (model) to the graphics card memory.
you have a loop to update the positions and draw the model using the graphics card memory.
i appreaciate your tutrials but i dont see how the memory can go that high as the model is in gpu memory unless there is a memory leak/bug/dependancy bug.
im going to be using the tutorial on a raspberry pi that has limited memory 

https://streamable.com/2b0rvo


@amengol be more constructive and post how to diagnoise properly instead of accusing people of having a 'nonsense issue'

**Comment by JoeyDeVries** [comment]
@jamieS95 The model will also be loaded in RAM as Assimp has an internal representation of the model. You can probably get rid of the memory by clearing Assimp's data the moment the mesh is stored in its VBO(s).  Even then, GPU drivers may still store VBO content/textures on RAM as things may not always fit GPU memory and things need to be swapped out based on draw calls. 

I'm not particularly proud of the model loading code and would def. structure things differently if I'd write it today, but it has done its job fine so far for helping people get some 'idea' of how to approach these topics. I still feel 400mb is quite large for the model so it wouldn't surprise me at all if something not right; maybe textures are loaded more than once still, or data is stored in duplicate locations? If you can figure that out, definitely sent me a pull request or a message so I'll fix things on my end.

**Comment by gameengineer** [comment]
@jamieS95  A issue title with a picture of memory usage from VS is far from being useful. It is better to do some work on your part to figure out what needs improving.   That being said I still contend there is no memory issue for this sample code that requires a "fix". Again to remind you this code is tutorial code and as such does not have the requirement of being efficient and highly performant. Real game engine code is written tight and often difficult to understand. It often is not ideal for teaching. If you need code that uses a small footprint that is up to you to write. Learn from this code base and try to engineer something for you own needs. Even if you did find memory improvements for Joey to include I do not think the memory usage for this tutorial code constitutes a code base issue that he needs to fix.
